### PR TITLE
remove grep color encodings when contents is not encoded

### LINKED ISSUE #7736: error when performing search + replace
I have a similar issue. The first two images show output for a search and replace that failed completely. The errors reported are rather cryptic. The overall count of successes vs fails seems totally random: 64 vs 11, when actually nothing seems to have been replaced. Replacement didn't happen in the first file, either, although it reports no error.

When I tried a really small and simple test with just two fake files and one word to be replaced, the whole replacing got stuck. See third image below.

Running version 1.3.1073 on Ubuntu 18.04.4

1
image
2
image
3
image

_Originally posted by @AltfunsMA in https://github.com/rstudio/rstudio/issues/2066#issuecomment-686894213_

### PR DESCRIPTION
**Pull request by melissa-barca**
### Intent

Fixes #7736 

If a user was outside of a project and didn't specify an encoding via Global Options, global replace would update files to include pieces of the color encodings from the grep command. used to find the pattern. 

### Approach

Some pieces of code were written under the impression that if no encoding was provided, there wasn't a need to decode the line. Decoding is not just for encodings different from UTF8, but is also used to parse the grep command.

### QA Notes

This specific issue occurred when the `Default text encoding` option was set to [Ask], but making changes to this code has a chance of affecting behavior on different operating systems. The changes here shouldn't affect the way global find/replace works for different encodings but it may be worth testing some anyways.

Note there is an existing issue for issues with Windows and regexes that has been around since 1.2 (maybe 1.1) being tracked here: https://github.com/rstudio/rstudio/issues/7012

### CODE REVIEW

File: src/cpp/session/modules/SessionFind.cpp [hunk 1]
```diff
@@ -1363,6 +1351,8 @@ core::Error runGrepOperation(const GrepOptions& grepOptions, const ReplaceOption
    std::string encoding = projects::projectContext().hasProject() ?
                           projects::projectContext().defaultEncoding() :
                           prefs::userPrefs().defaultEncoding();
+   if (encoding.empty())
+      encoding = "UTF-8";
```
Comment (jmcphers) [inline src/cpp/session/modules/SessionFind.cpp:1355]: Is this assumption valid on Windows?

File: src/cpp/session/modules/SessionFind.cpp [hunk 2]
```diff
@@ -1363,6 +1351,8 @@ core::Error runGrepOperation(const GrepOptions& grepOptions, const ReplaceOption
...
    std::string encodedString;
    error = r::util::iconvstr(grepOptions.searchPattern(),
```
Comment (jmcphers) [inline src/cpp/session/modules/SessionFind.cpp:1357]: Do we need to run `iconvstr` if the target encoding is UTF-8? (seems like asking it to convert UTF-8 to UTF-8 would be a noop)

File: src/cpp/session/modules/SessionFind.cpp (hunk 1 above)
Comment (melissa-barca) [reply src/cpp/session/modules/SessionFind.cpp:1355]: Yes, the only places we use the encoding are with `r::util::iconvstr` and the first thing that function does is change empty encodings to UTF-8. https://github.com/rstudio/rstudio/blob/master/src/cpp/r/RUtil.cpp#L143-L160

Comment (kevinushey) [reply src/cpp/session/modules/SessionFind.cpp:1355]: I don't think so -- on Windows, outside of an RStudio project, files will be saved in the system encoding, which will not be UTF-8. To that end, I think we need to be able to call R's `iconv()` machinery and tell it to conver from the native encoding to UTF-8. This is normally done by passing an empty string to `Riconv_open()`(https://cran.r-project.org/doc/manuals/r-release/R-exts.html#Re_002dencoding).

We do a strange here to check if the "effective" encodings are the same, to basically avoid unnecessary conversions:

https://github.com/rstudio/rstudio/blob/db6105f69c895b4c96d426ed95ef77bde14ff574/src/cpp/r/RUtil.cpp#L149-L160

but this check is also wrong (at least, on Windows; but also in theory on other platforms where a non-UTF8 locale has been set).

In short, I think:

1. It's okay to leave the encoding here as the empty string;
2. The shortcut in `iconvstr()` should avoid assuming that "empty string" -> "UTF-8".

File: src/cpp/session/modules/SessionFind.cpp (hunk 2 above)
Comment (melissa-barca) [reply src/cpp/session/modules/SessionFind.cpp:1357]: It's not a noop but possibly overkill, it sets `encodedString = grepOptions.searchPattern()` and will return early before attempting any conversion logic. `encodedString` writes the string in a temp file to be referenced later on.

File: src/cpp/session/modules/SessionFind.cpp (hunk 1 above)
Comment (melissa-barca) [reply src/cpp/session/modules/SessionFind.cpp]: I updated the code with Kevin's recommendation.

File: src/cpp/r/RUtil.cpp [hunk 3]
```diff
@@ -147,11 +147,7 @@ core::Error iconvstr(const std::string& value,
                      std::string* pResult)
 {
    std::string effectiveFrom = from;
-   if (effectiveFrom.empty())
```
Comment (jmcphers) [inline src/cpp/r/RUtil.cpp:150]: Even if we believe this change needs to be made, it dramatically increases the surface area of this patch since we call `iconvstr` all over the place. As this is a last-minute change before shipping, let's limit the changes to the find and replace codepath so that we can feel 100% confident we haven't broken or changed any other behavior.

**Review by jmcphers — CHANGES_REQUESTED** [review]

File: src/cpp/r/RUtil.cpp (hunk 3 above)
Comment (melissa-barca) [reply src/cpp/r/RUtil.cpp]: That's a good point. Change has been reverted; I still removed the lines where I was setting empty to "UTF-8" in `SessionFind` as it doesn't change any logic and is inaccurate.

Comment (jmcphers) [reply src/cpp/r/RUtil.cpp]: 👍

**Review by jmcphers — APPROVED** [review]

**Comment by jmcphers** [comment]
@kevinushey we will wait for your explicit approval before merging this (want to get at least 2 sets of eyes on it since we're slamming it in right before shipping!)

**Review by kevinushey — APPROVED** [review]
LGTM!

**Comment by R-Hannibal** [comment]
Adding some additional QA Notes for myself after talking with @melissa-barca:

-We do not need to do a full check of find and replace functionality to verify this fix, but due to the `Default text encoding` option code being slightly different between different OS, we should do basic tests across multiple OS using different encoding options - Global Options > Code > Saving. We don't need to tests EVERY Default text encoding option, but should at least hit a couple of different ones per OS.
-This has already been tested in Dev environments, but I'd like to see QA check it in AWS server instances as well (we already have them set up for 1.3 certification, shouldn't be too big a pain).

image (1)
